services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/CompletedDirectoryDownload.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.transfer.s3;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+import software.amazon.awssdk.annotations.SdkPreviewApi;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Represents a completed download directory transfer to Amazon S3. It can be used to track
+ * failed single file downloads.
+ *
+ * @see S3TransferManager#downloadDirectory(DownloadDirectoryRequest) 
+ */
+@SdkPublicApi
+@SdkPreviewApi
+public final class CompletedDirectoryDownload implements CompletedDirectoryTransfer {
+
+    private final Collection<FailedFileDownload> failedTransfers;
+
+    private CompletedDirectoryDownload(DefaultBuilder builder) {
+        this.failedTransfers = Collections.unmodifiableCollection(
+            Validate.paramNotNull(builder.failedTransfers, "failedTransfers"));
```

Comment:
Defensive copy in case of builder reuse?

Reply:
We should also probably treat failedTransfers == null as being empty. CompletedDirectoryDownload.builder().build() should probably be valid.

Reply:
The array is copied in the builder setter

Reply:
My concern was that this behavior seems possible, unless I'm misreading:

```java
CompletedDirectoryDownload.Builder builder = CompletedDirectoryDownload.builder().addFailedTransfer(foo);
CompletedDirectoryDownload download = builder.build();

System.out.println(download.failedTransfers().size()); // Prints 1
builder.addFailedTransfer(foo);
System.out.println(download.failedTransfers().size()); // Prints 2
```

Reply:
Ah yeah, good catch